pkg/credentialwatcher/secretwatcher.go

```diff
@@ -0,0 +1,129 @@
+var SecretWatcher *secretWatcher
+
+type secretWatcher struct {
+	watchInterval time.Duration
+	client        client.Client
+}
+
+// Initialize creates a global instance of the SecretWatcher
+func Initialize(client client.Client, watchInterval time.Duration) {
+	SecretWatcher = NewSecretWatcher(client, watchInterval)
+}
+
+// NewSecretWatcher returns a new instance of the SecretWatcher interface
+func NewSecretWatcher(client client.Client, watchInterval time.Duration) *secretWatcher {
+	return &secretWatcher{
+		watchInterval: watchInterval,
+		client:        client,
+	}
+}
+
+// SecretWatcher will trigger CredentialsRotator every `scanInternal` and only stop if the operator is killed or a
+// message is sent on the stopCh
+func (s *secretWatcher) Start(log logr.Logger, stopCh <-chan struct{}) {
+	log.Info("Starting the secretWatcher")
+	for {
+		select {
+		case <-time.After(s.watchInterval):
+			log.Info("secretWatcher: scanning secrets")
+			err := s.ScanSecrets(log)
+			if err != nil {
+				log.Error(err, "secretWatcher not started, credentials wont be rotated")
+			}
+		case <-stopCh:
+			log.Info("Stopping the secretWatcher")
+		}
+	}
+}
+
+// CredentialsRotator will list all secrets with the `STSCredentialsSuffix` and mark the account CR `status.rotateCredentials` true
+// if the credentials CreationTimeStamp is within `STSCredentialsRefreshThreshold` of `STSCredentialsDuration`
+func (s *secretWatcher) ScanSecrets(log logr.Logger) error {
+	// List STS secrets and check their expiry
+	secretList := &corev1.SecretList{}
+
+	listOps := &client.ListOptions{Namespace: awsv1alpha1.AccountCrNamespace}
+	if err := s.client.List(context.TODO(), listOps, secretList); err != nil {
+		log.Error(err, fmt.Sprintf("Unable to list secrets in namespace %s", awsv1alpha1.AccountCrNamespace))
+		return err
+	}
+
+	for _, secret := range secretList.Items {
+
+		accountName := strings.TrimSuffix(secret.ObjectMeta.Name, STSCredentialsSuffix)
+
+		if strings.HasSuffix(secret.ObjectMeta.Name, STSCredentialsSuffix) {
+			unixTime := time.Unix(secret.ObjectMeta.CreationTimestamp.Unix(), 0)
+			timeSinceCreation := int(time.Since(unixTime).Seconds())
+
+			if STSCredentialsDuration-timeSinceCreation < STSCredentialsThreshold {
+
+				accountInstance, err := s.GetAccount(accountName)
+				if err != nil {
+					log.Error(err, "Unable to rerieve account CR %s", accountName)
+				}
+
+				if accountInstance.Status.RotateCredentials != true {
+
+					log.Info(fmt.Sprintf("AWS credentials secret %s was created %s ago requeing to be refreshed", secret.ObjectMeta.Name, time.Since(unixTime)))
+
+					accountInstance.Status.RotateCredentials = true
+
+					err = s.UpdateAccount(accountInstance)
+					if err != nil {
+						log.Error(err, fmt.Sprintf("Error updating account %s", accountName))
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// GetAccount retrieve account CR
+func (s *secretWatcher) GetAccount(accountName string) (*awsv1alpha1.Account, error) {
```

Comment:
this function only saves 1 line of code at 86 not sure it is necessary.

Reply:
It's keeping the logic separate, I didn't want `ScanSecrets()` to become too big. Even as it is I think it should be cleaned up a bit, specially the time comparison part.